File: Algorithms/Java/Populating Next Right Pointers in Each Node II.java
/**
 * Definition for binary tree with next temper.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void connect(TreeLinkNode root) {
        if (root == null)
			return;
		TreeLinkNode temp = root.next;
		while (temp != null) {
			if (temp.left != null) {
				temp = temp.left;
				break;
			} else if (temp.right != null) {
				temp = temp.right;
				break;
			} else
				temp = temp.next;
		}
		if (root.right != null) {
			root.right.next = temp;
		}
		if (root.left != null) {
			root.left.next = root.right == null ? temp : root.right;
		}
		connect(root.right);
		connect(root.left);
    }
}

File: Algorithms/Python/Populating-Next-Right-Pointers-in-Each-Node-II.py
# Definition for binary tree with next pointer.
# class TreeLinkNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

# METHOD 1: Space: O(1), Time: O(n) --- BFS


class Solution(object):

    def connect(self, root):
        """
        :type root: TreeLinkNode
        :rtype: nothing
        """
        head = root
        while head:
            pre, cur, nex = None, head, None
            while cur:
                if not nex:
                    nex = cur.left if cur.left else cur.right
                if cur.left:
                    if pre:
                        pre.next = cur.left
                    pre = cur.left
                if cur.right:
                    if pre:
                        pre.next = cur.right
                    pre = cur.right
                cur = cur.next
            head = nex

# METHOD 2: Space: O(logN), Time: O(n) --- DFS


class Solution(object):

    def connect(self, root):
        """
        :type root: TreeLinkNode
        :rtype: nothing
        """
        if not root:
            return
        temp = root.next
        while temp:
            if temp.left:
                temp = temp.left
                break
            if temp.right:
                temp = temp.right
                break
            temp = temp.next
        if root.right:
            root.right.next = temp
        if root.left:
            root.left.next = root.right if root.right else temp
        self.connect(root.right)
        self.connect(root.left)
